[Chen Xi] Duke Increments

chen-xi-cx opened the pull request:
Add week 2 and week 3 iP increments

j-lum committed:
Add support for Gradle workflow

j-lum committed:
Add sample checkstyle configuration

Add toolVersion block in to Gradle code sample to prevent errors.

j-lum committed:
Change file mode on `gradle` to be executable

j-lum committed:
Merge pull request #12 from j-lum/gradle+x

Change file mode on `gradle` to be executable (#9)

j-lum committed:
Add configuration for console applications

Gradle defaults to an empty stdin which results in runtime exceptions
when attempting to read from `System.in`. Let's add some sensible
defaults for students who may still need to work with the standard
input stream.

j-lum committed:
Merge pull request #13 from j-lum/javaexec

Add configuration for console applications

chen-xi-cx committed:
Add Level-1 increment

chen-xi-cx committed:
fix syntax of Duke constructor

chen-xi-cx committed:
Add Level-2 increment

chen-xi-cx committed:
Add Task.java and Level-3 increment

chen-xi-cx committed:
Add Deadline.java

chen-xi-cx committed:
add Event.java

chen-xi-cx committed:
Add ToDo.java

chen-xi-cx committed:
Add Level-4 increment

chen-xi-cx committed:
Add DukeException.java

chen-xi-cx committed:
Add Level-5 increment.Handle error: todo/event/deadline without detail/date.

chen-xi-cx committed:
Add Level-6 increment: Duke now supports delete task function

chen-xi-cx committed:
Duke.java: fix indentation of switch case

chen-xi-cx committed:
Add Level-7 increment: duke can save and read task list from hard disk

chen-xi-cx committed:
Add Level-8 increment: store date as Date type instead of String

chen-xi-cx committed:
Merge branch 'branch-Level-7'

chen-xi-cx committed:
Fix merge conflict in Duke.java

chen-xi-cx committed:
Fix Level-8 increment bug after merging branch-Level-8

chen-xi-cx committed:
Remove StringToDateConverter.java

chen-xi-cx committed:
Add part of More OOP: add loading files from Storage Class

chen-xi-cx committed:
Add more OOP: refactor bye and done command

chen-xi-cx committed:
Add more OOP: refactor Delete command

chen-xi-cx committed:
Add more OOP: refactor Todo command

remove Deadline and Event command

chen-xi-cx committed:
Add more OOP: refactor Deadline and Event command

chen-xi-cx committed:
Remove unused imports

chen-xi-cx committed:
Add more OOP: refactor file saving feature

Fix incorrect loading of ToDo from duke.txt

chen-xi-cx committed:
Remove TaskFileReader.java

chen-xi-cx committed:
Remove TaskFileWriter.java

chen-xi-cx committed:
Add more OOP: refactor DukeException

Add A-MoreOOP increment

chen-xi-cx committed:
Add data directory

chen-xi-cx committed:
Remove duke.txt

chen-xi-cx committed:
Add A-Packages increment

chen-xi-cx committed:
Add A-JUnit increment

chen-xi-cx committed:
Add A-Jar increment

chen-xi-cx committed:
Add JavaDoc to Command.java

chen-xi-cx committed:
Add A-JavaDoc: add JavaDoc comment to all files

chen-xi-cx committed:
Add A-CodingStandard: fix all code to follow coding standard

chen-xi-cx committed:
Add Level-9 increment: add find command

chen-xi-cx committed:
Merge branch 'branch-A-JavaDoc'
Add A-JavaDoc increment

chen-xi-cx committed:
Merge branch 'branch-A-CodingStandard'
Add A-CodingStandard increment

chen-xi-cx committed:
Merge branch 'branch-Level-9'
Add Level-9 increment to master

liawsy commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    /**
+     * Execute a command.
```
Minor issue, should be "Executes a command", refer to Javadocs section in [Java coding standards](https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html).

liawsy commented on src/main/java/converter/StringDateConverter.java:
```diff
@@ -0,0 +1,35 @@
+package converter;
+
+import java.text.SimpleDateFormat;
```
Cool use of Java built-in classes!

liawsy commented on src/main/java/exception/DukeException.java:
```diff
@@ -0,0 +1,35 @@
+        switch(this.type) {
+        case "invalid number":
+            return "Please enter a valid number.";
+        case "empty todo":
```
error message type can be more concise.

liawsy commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    public Command parse(String userInput) throws DukeException {
+        String[] separatedInputs = userInput.trim().split("\\s+");
+        switch (separatedInputs[0].toLowerCase()) {
```
nice consideration of how input could possibly be capitalised.

liawsy reviewed (commented):
Clean and readable code. Also, good organisation of your code using the packages. Here're some other feedback:
1) Javadocs are written in a detailed and informative manner, although be sure to note that by standards, each statement should end with a full stop.
2) Instead of using the `IncorrectCommand` class, you could make it an exception.
3) Perhaps you could create a class that handles all the messages, so that when you want to make changes to the messages in the UI, you don't have to go through every single file that contains their respective messages to make changes. 

Overall, really good job and I enjoyed reading your code.

chen-xi-cx committed:
Merge branch 'gradle'

chen-xi-cx committed:
Modify build.gralde: change mainClassName

Fix done and delete command index out of bound error when user input is 'done' or 'delete'

chen-xi-cx committed:
Add check style and resolve all check style error

chen-xi-cx committed:
Add A-Gradle increment

chen-xi-cx committed:
Merge branch 'branch-A-Gradle'

chen-xi-cx committed:
Modify build.gradle to support JavaFx

aidilfbk commented on src/main/java/converter/StringDateConverter.java:
```diff
@@ -0,0 +1,35 @@
+/**
+ * Converts String object to Date object.
+ */
+public class StringDateConverter {
```
I think it would be a good idea to turn the methods in this class to be static as they do not seem to be using any instance specific properties.

aidilfbk commented on src/main/java/converter/StringDateConverter.java:
```diff
@@ -0,0 +1,35 @@
+     * @return date object
+     * @throws Exception if the date is unparesable
+     */
+    public Date convertStringToDate(String strDate) throws Exception {
```
Good use of checked exceptions but I think it could provide more clarity by being more specific with the type of exception that may be thrown.

aidilfbk commented on src/main/java/exception/DukeException.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @Override
+    public String getMessage() {
+        switch (this.type) {
```
This would be a good place to use enums so other developers don't have to remember the specific strings that correspond to a specific error message.

aidilfbk commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,78 @@
+     * @return the command based on user input
+     * @throws DukeException if user input is invalid
+     */
+    public Command parse(String userInput) throws DukeException {
```
I think it would be a good idea to make the `parse` method static as it does not seem to be using any instance specific properties.

aidilfbk commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,78 @@
+                throw new DukeException("invalid number");
+            }
+            if (separatedInputs[0].toLowerCase().equals("done")) {
+                return new DoneCommand(s.nextInt() - 1);
```
I may be misunderstanding the condition that will trigger this line of code. The outer switch checks for `separatedInputs[0].toLowerCase()` being equal to `"delete"` but there is also an if-else conditional check for `separatedInputs[0].toLowerCase().equals("done")`.

I believe these two cases are mutually exclusive but is there an example input test case that explains it better?

aidilfbk commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,115 @@
+     * @return list of task if the file is found
+     * @throws FileNotFoundException when no task list file is found
+     */
+    public ArrayList<Task> load() throws FileNotFoundException {
```
Clear logic of how the file contents is parsed. :+1:

aidilfbk commented on src/main/java/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,48 @@
+        try {
+            StringDateConverter converter = new StringDateConverter();
+            Date by = converter.convertStringToDate(arguments[1].trim());
+            tasks.getTasks().add(new Deadline(arguments[0], by));
```
I think it would provide clarity if there was a temporary variable to reference the `new Deadline` instance since it is later used, rather than requesting it again from the `tasks` variable.

A similar issue appears in [`DoneCommand`](https://github.com/nus-cs2103-AY1920S1/duke/pull/187/files#diff-a1acc0716924f84df197496c2462f84fR36).

chen-xi-cx committed:
Add GUI for Duke

GUI now shows welcome message but Duke only reply haha to any input

chen-xi-cx committed:
Refactor all Command class and Ui class

chen-xi-cx committed:
Add Level-10 increment: Duke now has GUI

chen-xi-cx committed:
Merge branch 'branch-Level-10'

aidilfbk commented on src/main/java/exception/DukeException.java:
```diff
@@ -0,0 +1,37 @@
+
+    /**
+     * Initializes DukeException with error type.
+     * @param input the error type
```
Minor Javadoc style violation as there should be an [empty line between description and parameter section](https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html#comments). This style is adhered to in other Javadoc comments so I believe this style violation was unintended.

aidilfbk commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,30 @@
+        this.tasks = tasks;
+    }
+
+    public ArrayList<Task> getTasks() {
```
It seems like every other class that uses a `TaskList` needs to call this method to do their operations. Consider providing those methods (e.g. `add(Task)`) on the `TaskList` instead of allowing unrestricted manipulation of the `ArrayList<Task>`.

aidilfbk commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,78 @@
+        case "done":
+            //fallthrough
+        case "delete":
+            if (separatedInputs.length <= 1) {
```
I like that input validation is performed as early as possible so that `Command` implementations can be simplified. 👍

aidilfbk reviewed (commented):
Hi Chen Xi,

I've made my comments on specific lines where relevant and my thoughts in general are below:
- Very good use of packaging :+1:
- Code is well organized and logically implemented, making it easy to follow.
- I'm not sure if the Javadoc style for parameters should treat the parameter name as part of the description (e.g. "[@param tasks contains task list](https://github.com/nus-cs2103-AY1920S1/duke/pull/187/files#diff-ae04d255fd2aa7bba6aee1540703424eR23)") but I'll check with my TA on this matter.
- I think it would provide more clarity if method implementations were logically chunked (see ["logical units within a block should be separated by one blank line" from the Coding Standard](https://nus-cs2103-ay1920s1.github.io/website/coding-standards/java/intermediate.html#layout)).

I hope you found my feedback useful! 😄

chen-xi-cx committed:
Add Java assert to DeleteCommand.java and DoneCommand.java

chen-xi-cx committed:
Merge pull request #4 from chen-xi-cx/branch-A-Assertions

Add A-Assertions increment

chen-xi-cx committed:
Rewrite exception logic of StringDateConverter.java

chen-xi-cx committed:
Modify EventCommand.java to catch ParseException

chen-xi-cx committed:
Rewrite Exception logic of DeleteCommand.java and DoneCommand.java

chen-xi-cx committed:
Rewrite logic of DukeException.java to use Java enum

chen-xi-cx committed:
Refactor save() method in Storage.java

chen-xi-cx committed:
Rename attribute in DeadlineCommand.java and DeleteCommand.java

Delete IncorrectCommand.java

chen-xi-cx committed:
Refactor Parser.java

chen-xi-cx committed:
Rework StringDateConverter.java logic to reject invalid date

chen-xi-cx committed:
Modify return statement of isValidStrDate in StringDateConverter.java

chen-xi-cx committed:
Refactor Ui.java: ensure SLAP

Fix add event command not working

chen-xi-cx committed:
Merge pull request #5 from chen-xi-cx/branch-A-CodeQuality

Add Branch-A-CodeQuality

chen-xi-cx committed:
Rename enum in DukeException.java

chen-xi-cx committed:
Fix Duke not showing printing exception message

Fix IndexOutOfBoundException when command is '/at' or '/by' with no date supplied

chen-xi-cx committed:
Merge pull request #6 from chen-xi-cx/branch-A-CodeQuality

Fix Duke not showing printing exception message

chen-xi-cx committed:
Merge branch 'master' of https://github.com/chen-xi-cx/duke

chen-xi-cx committed:
Add edit feature: to edit a task, enter 'edit [task number] [task field] [new date or description]'

Remove TaskTest.java

chen-xi-cx committed:
Merge pull request #7 from chen-xi-cx/branch-C-Update

Add edit feature

chen-xi-cx committed:
Modify dialog box to extend in height to accomodate long text

chen-xi-cx committed:
GUI closes automatically after bye

chen-xi-cx committed:
Change Duke dialogbox to light grey

Remove top spacing of dialogbox

chen-xi-cx committed:
Add spacing between children in dialogbox

chen-xi-cx committed:
Change chatbot and user profile picture

chen-xi-cx committed:
Change mainclassname to Launcher in build.gradle

Change display message when storage cannot find file to load

chen-xi-cx committed:
Add data directory to store task list

chen-xi-cx committed:
Change invalid date error message

chen-xi-cx committed:
Add title to GUI window

chen-xi-cx committed:
Simplify displayed date on Duke GUI

Modify display message of done command and delete command

chen-xi-cx committed:
Create new data directory if it does not exist

chen-xi-cx committed:
Remove duke.txt

chen-xi-cx committed:
Fix grammar error in display message

Change DaUser picture

Change dialog box padding

chen-xi-cx committed:
Add User Guide

chen-xi-cx committed:
Add Ui.png

Modify user guide

chen-xi-cx committed:
Fix typo in user guide

chen-xi-cx committed:
Fix error in gradle test

Change jar version to v0.1 in build.gralde

chen-xi-cx committed:
Add new features to README.md

chen-xi-cx committed:
Add cross-platform dependencies in build.gradle

chen-xi-cx committed:
Moves Launcher.java to main directory

Change group to main in build.gradle

chen-xi-cx committed:
Remove redundant code

chen-xi-cx committed:
Finalize everything. Fix cross-platform issues

chen-xi-cx committed:
Ready for release

chen-xi-cx committed:
Code cleanup
